browser/src/Services/VersionControl/VersionControlBlameLayer.tsx

```diff
@@ -0,0 +1,386 @@
+// CursorBufferLine - The 1 based position of the cursor in the file i.e. at line 30 it will be 30
+// CursorScreenLine - the position of the cursor within the visible lines so if line 30 is at the
+// top of the viewport it will be 0
+
+export class Blame extends React.PureComponent<IProps, IState> {
+    // Reset show blame to false when props change - do it here so it happens before rendering
+    // hide if the current line has changed or if the text of the line has changed
+    // aka input is in progress or if there is an empty line
+    public static getDerivedStateFromProps(nextProps: IProps, prevState: IState) {
+        const lineNumberChanged = nextProps.cursorBufferLine !== prevState.currentCursorBufferLine
+        const lineContentChanged = prevState.currentLineContent !== nextProps.currentLine
+        if (
+            (prevState.showBlame && (lineNumberChanged || lineContentChanged)) ||
+            !nextProps.currentLine
+        ) {
+            return {
+                showBlame: false,
+                blame: prevState.blame,
+                currentLineContent: nextProps.currentLine,
+                currentCursorBufferLine: nextProps.cursorBufferLine,
+            }
+        }
+        return null
+    }
+
+    public state: IState = {
+        error: null,
+        blame: null,
+        showBlame: null,
+        currentLineContent: this.props.currentLine,
+        currentCursorBufferLine: this.props.cursorBufferLine,
+    }
+
+    private _timeout: any
+    private readonly DURATION = 300
+    private readonly LEFT_OFFSET = 4
+
+    public async componentDidMount() {
+        const { cursorBufferLine, mode } = this.props
+        await this.updateBlame(cursorBufferLine, cursorBufferLine)
+        if (mode === "auto") {
+            this.resetTimer()
+        }
+        this.props.setupCommand(() => {
+            const { showBlame } = this.state
+            this.setState({ showBlame: !showBlame })
+        })
+    }
+
+    public async componentDidUpdate(prevProps: IProps, prevState: IState) {
+        const { cursorBufferLine, currentLine, mode } = this.props
+        if (prevProps.cursorBufferLine !== cursorBufferLine && currentLine) {
+            await this.updateBlame(cursorBufferLine, cursorBufferLine)
+            if (mode === "auto") {
+                return this.resetTimer()
+            }
+        }
+    }
+
+    public componentWillUnmount() {
+        clearTimeout(this._timeout)
+    }
+
+    public componentDidCatch(error: Error) {
+        warn(`Oni VCS Blame layer failed because: ${error.message}`)
+        this.setState({ error })
+    }
+
+    public resetTimer = () => {
+        clearTimeout(this._timeout)
+        this._timeout = setTimeout(() => {
+            if (this.props.currentLine) {
+                this.setState({ showBlame: true })
+            }
+        }, this.props.timeout)
+    }
+
+    public getLastEmptyLine() {
+        const { cursorLine, visibleLines, topBufferLine } = this.props
```

Comment:
Nice work getting all this info from the buffer layer API, it's another great example 💯